feat(python): add OpenOffice spreadsheet support via `read_ods`

alexander-beedie opened the pull request:
I didn't like having the following in the docstring...
```
If read_excel does not work or you need to read other
types of spreadsheet files, you can try pandas pd.read_excel()
```
...so I bit the bullet and have added support for parsing OpenOffice `.ods` spreadsheet files (and removed that notice). 

## Performance

Benchmarking indicates that we'll be a little over _**5x** faster_ than pandas at parsing the same data. Testing with ½ million elements of mixed-type table data (str, int, float, date, datetime) resulted in the following timings:
```python
import pandas as pd
with Timer():
    pf = pd.read_ods("data.ods")
    # Elapsed time: 15.7151 seconds
```
```python
import polars as pl
with Timer():
    df = pd.read_ods("data.ods")
    # Elapsed time: 2.9356 seconds 🚀
 ```

## Dependencies

We take on _no new dependencies_ (except for unit testing). As usual, if the caller doesn't already have the necessary modules in their environment we raise an `ImportError` and tell them what they need to `pip install`.

## Follow-up

Now that we have three engines I think we should probably lift-up (as first-class parameters) one or two features that are currently only available via `xlsx2csv` with `read_csv_options`; only considering "has_headers" at the moment (so that we can load headerless sheet/table data using the other engines too), as this is a fairly common/fundamental option. Will leave any such changes here for a subsequent PR.

## Also

Added support for the `schema_overrides` param (as we have elsewhere, such as `read_database`, frame init, etc) so that the caller can fine-tune the loaded `DataFrame` column dtypes. Additional minor tidying/cleanups.

alexander-beedie commented:
Rebased to pickup "boolean trap" updates.

stinodego commented:
This is great stuff! I'm taking a closer look tomorrow.

stinodego reviewed (changes requested):
Great to have ODS parsing functionality! However, why is this part of `read_excel`? It's not an Excel file. Maybe we should rebrand `read_excel` to `read_spreadsheet`, or add a separate function specifically for ODS files.

Otherwise, the implementation looks good! I left a few minor comments.

, but my main question is: why is this part of the `read_excel` function? It's not an Excel file.

alexander-beedie commented:
> my main question is: why is this part of the `read_excel` function? It's not an Excel file.

Actually Excel can work with ODS files too, so it kind of is ;)

We definitely don't want two functions for this. I was thinking about `read_spreadsheet`, but we then lose symmetry with `write_excel`, and I don't think the leap from "I have a spreadsheet" to "I can probably read it with `read_excel`" is a large one. I think keeping the name as-is seems reasonable.

alexander-beedie commented on py-polars/polars/io/excel/functions.py:
```diff
@@ -249,70 +267,196 @@ def read_excel(
         try:
-            import xlsx2csv
+            import ezodf
         except ImportError:
-            raise ModuleNotFoundError(
-                "xlsx2csv is not installed\n\nPlease run: `pip install xlsx2csv`"
+            raise ImportError(
+                "Required package not installed\n\nPlease run `pip install ezodf lxml`"
             ) from None
```
We have used `dependencies` for core functionality that we think is likely to be loaded, but we want to defer the first load in case it isn't. Less usual/edge-case functionality that we think is unlikely to be used has tended to be import-guarded instead.

stinodego replied:
I'll make an issue for determining what to do with optional dependencies. Doesn't have to block this PR.

alexander-beedie commented:
Update: if you do think we're better distinguishing more clearly between OpenOffice/Excel then (which is growing on me, heh), I suggest we *do* actually go for two functions.

* `read_excel` (excel, obviously ;)
* `read_odf` (open document format)

This would give us clarity in the read functions AND potential future symmetry should we ever find that we want a `write_odf`.

stinodego commented:
> (Update2: This option is growing on me - I'll split it out into a read_ods tomorrow ;)

That sounds good to me!

alexander-beedie commented:
> That sounds good to me!

Done! Refactored it into two distinct methods, sharing common code via a private `_read_spreadsheet` function. Renamed the `io/excel` subdirectory as `io/spreadsheet`, and added some more test coverage.

stinodego reviewed (commented):
Looks great! I had one comment on the tests, otherwise good to go in my opinion.

alexander-beedie commented on py-polars/tests/unit/io/test_spreadsheet.py:
```diff
@@ -14,38 +15,79 @@
 @pytest.fixture()
-def excel_file_path(io_files_path: Path) -> Path:
-    return io_files_path / "example.xlsx"
+def excel_file_path(io_files_path: Path) -> str:
```
> I'd rather work with `Path`s than with strings - `Path` objects should be the default. Is there a reason you changed these around?

Yup; use of `request.getfixturevalue` (which allowed me to parametrize the xls/ods paths) requires a string. Can always call the `str(path)` in the `getfixturevalue` call if you prefer, but given the usage it seemed neater to just have the fixtures return strings here.

stinodego replied:
All right - that `getfixturevalue` feels a bit high-tech for my tastes. Tests should be dumb and simple. But I'll allow it :smile:

alexander-beedie commented:
(Rebased to resolve `requirements` conflict)

stinodego commented:
All right, no reason not to press that green button, right?
